src/main/java/org/jitsi/videobridge/cc/BitrateController.java

```diff
@@ -927,27 +1004,47 @@ private synchronized void update()
             logger.debug("Prioritizing endpoints, adjusted last-n: " + adjustedLastN +
                 ", sorted endpoint list: " +
                 conferenceEndpoints.stream().map(AbstractEndpoint::getID).collect(Collectors.joining(", ")) +
-                ". Selected endpoints: " + String.join(", ", selectedEndpointIds) +
-                ". Pinned endpoints: " + String.join(", ", pinnedEndpointIds));
+                ". Endpoints constraints: " + Arrays.toString(endpointConstraintsMap.values().toArray()) +
+                ". Global endpoint constraints" + globalConstraints);
         }
 
         int endpointPriority = 0;
 
-        // First, bubble-up the selected endpoints (whoever's on-stage needs to
-        // be visible).
-        for (Iterator<AbstractEndpoint> it = conferenceEndpoints.iterator();
-             it.hasNext() && endpointPriority < adjustedLastN;)
+        // Before endpoint constraints we had had the notion of selected and
+        // pinned endpoints. "selected" endpoints were endpoints that had
+        // 720p ideal height and prioritized over everything else and were
+        // allocated bandwidth first (up to their preferred resolution).
+        // "Pinned" endpoints where those thee had 180p ideal height and were
+        // ranked lower than selected endpoints but before the rest of the pack.
+        // Finally, any remaining endpoints were ranked after pinned endpoints.
+
+        Map<String, EndpointConstraints> endpointConstraintsMapCopy = endpointConstraintsMap;
+        EndpointConstraints globalConstraintsCopy = globalConstraints;
+
+        List<EndpointMultiRank> endpointMultiRankList = IntStream
+            .range(0, conferenceEndpoints.size() - 1)
+            .mapToObj(i -> {
+                AbstractEndpoint endpoint = conferenceEndpoints.get(i);
+                EndpointConstraints endpointConstraints = globalConstraintsCopy
+                    .of(endpoint.getID())
+                    .unless(endpointConstraintsMapCopy.get(endpoint.getID()));
```

Comment:
So we translate the 'global constraints' into instances of Endpoint constraints for every endpoint (that doesn't already have a specific constraint set)?

Reply:
Yes, that was the thinking. But we no longer need to translate now that we have `VideoConstraints`.